File: src/main/java/com/mineinabyss/pyroscope/PyroscopePlugin.java
package com.mineinabyss.pyroscope;

import com.mojang.brigadier.Command;
import io.papermc.paper.command.brigadier.Commands;
import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvents;
import io.pyroscope.http.Format;
import io.pyroscope.javaagent.EventType;
import io.pyroscope.javaagent.PyroscopeAgent;
import io.pyroscope.javaagent.config.Config;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.HashMap;
import java.util.Map;

public class PyroscopePlugin extends JavaPlugin {
    @SuppressWarnings("UnstableApiUsage")
    @Override
    public void onEnable() {
        saveDefaultConfig();
        start();
        getLifecycleManager().registerEventHandler(LifecycleEvents.COMMANDS, event -> event.registrar().register(
                Commands.literal("pyroscope")
                        .requires(it -> it.getSender().hasPermission("pyroscope.admin"))
                        .then(Commands.literal("reload").executes(ctx -> {
                            stop();
                            start();
                            ctx.getSource().getSender().sendMessage("Pyroscope reloaded");
                            return Command.SINGLE_SUCCESS;
                        })).build()
        ));
    }

    private void start() {
        reloadConfig();
        var env = System.getenv();
        var config = getConfig();
        Map<String, String> labels = config.getObject("labels", Map.class, new HashMap());
        labels.putIfAbsent("server", env.getOrDefault("SERVER_NAME", "unknown"));
        getLogger().info("Starting Pyroscope");
        PyroscopeAgent.start(new Config.Builder()
                .setLabels(labels)
                .setApplicationName(config.getString("applicationName"))
                .setProfilingEvent(EventType.ITIMER)
                .setProfilingAlloc(config.getString("allocSize"))
                .setFormat(Format.JFR)
                .setServerAddress(config.getString("serverAddress"))
                .build()
        );
    }

    @Override
    public void onDisable() {
        getLogger().info("Stopping Pyroscope");
        stop();
    }

    private void stop() {
        PyroscopeAgent.stop();
    }
}

File: src/main/kotlin/com/mineinabyss/pyroscope/PyroscopePlugin.kt
package com.mineinabyss.pyroscope

import com.mineinabyss.idofront.commands.brigadier.commands
import com.mineinabyss.idofront.config.config
import io.pyroscope.http.Format
import io.pyroscope.javaagent.EventType
import io.pyroscope.javaagent.PyroscopeAgent
import kotlinx.serialization.Serializable
import org.bukkit.plugin.java.JavaPlugin
import io.pyroscope.javaagent.config.Config as PyroscopeConfig

class PyroscopePlugin : JavaPlugin() {
    @Serializable
    data class Config(
        val applicationName: String = "minecraft",
        val serverAddress: String = "http://pyroscope-server:4040",
    )

    override fun onEnable() {
        start()
        commands {
            "pyroscope" {
                requiresPermission("pyroscope.admin")
                "reload" {
                    executes {
                        stop()
                        start()
                        sender.sendMessage("Pyroscope reloaded")
                    }
                }
            }
        }
    }

    private fun start() {
        val env = System.getenv()
        val labels = mapOf("server" to (env["SERVER_NAME"] ?: "unknown"))
        val config by config("config", dataPath, Config())
        logger.info("Starting Pyroscope with config: $config")
        PyroscopeAgent.start(
            PyroscopeConfig.Builder()
                .setLabels(labels)
                .setApplicationName(config.applicationName)
                .setProfilingEvent(EventType.ITIMER)
                .setProfilingAlloc("512k")
                .setFormat(Format.JFR)
                .setServerAddress(config.serverAddress)
                .build()
        )
    }

    override fun onDisable() {
        logger.info("Stopping Pyroscope")
        stop()
    }

    private fun stop() {
        PyroscopeAgent.stop()
    }
}
